x/proof/keeper/validate_proofs.go

```diff
@@ -0,0 +1,206 @@
+// It removes them from the store once they are processed.
+func (k Keeper) ValidateSubmittedProofs(ctx sdk.Context) (numValidProofs, numInvalidProofs uint64, err error) {
+	logger := k.Logger().With("method", "ValidateSubmittedProofs")
+
+	logger.Info(fmt.Sprintf("Number of CPU cores used for parallel proof validation: %d\n", numCPU))
+
+	// Iterate over proofs using an proofIterator to prevent memory issues from bulk fetching.
+	proofIterator := k.GetAllProofsIterator(ctx)
+
+	coordinator := &proofValidationTaskCoordinator{
+		// Parallelize proof validation across CPU cores since they are independent from one another.
+		// Use semaphores to limit concurrent goroutines and prevent memory issues.
+		sem: make(chan struct{}, numCPU),
+		// Use a wait group to wait for all goroutines to finish before returning.
+		wg: &sync.WaitGroup{},
+
+		processedProofs: make(map[string][]string),
+		coordinatorMu:   &sync.Mutex{},
+	}
+
+	for ; proofIterator.Valid(); proofIterator.Next() {
+		proofBz := proofIterator.Value()
+
+		// Acquire a semaphore to limit the number of goroutines.
+		// This will block if the sem channel is full.
+		coordinator.sem <- struct{}{}
+
+		// Increment the wait group to wait for proof validation to finish.
+		coordinator.wg.Add(1)
+
+		go k.validateProof(ctx, proofBz, coordinator)
+	}
+
+	// Wait for all goroutines to finish before returning.
+	coordinator.wg.Wait()
+
+	// Close the proof iterator before deleting the processed proofs.
+	proofIterator.Close()
+
+	// Delete all the processed proofs from the store since they are no longer needed.
+	logger.Info("removing processed proofs from the store")
+	for supplierOperatorAddr, processedProofs := range coordinator.processedProofs {
+		for _, sessionId := range processedProofs {
+			k.RemoveProof(ctx, sessionId, supplierOperatorAddr)
+			logger.Info(fmt.Sprintf(
+				"removing proof for supplier %s with session ID %s",
+				supplierOperatorAddr,
+				sessionId,
+			))
+		}
+	}
+
+	return coordinator.numValidProofs, coordinator.numInvalidProofs, nil
+}
+
+// validateProof validates a proof before removing it from the store.
+// It marks the corresponding claim as valid or invalid based on the proof validation.
+// It is meant to be called concurrently by multiple goroutines to parallelize
+// proof validation.
+func (k Keeper) validateProof(
+	ctx context.Context,
+	proofBz []byte,
+	coordinator *proofValidationTaskCoordinator,
+) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	logger := k.Logger().With("method", "validateProof")
+
+	// Decrement the wait group when the goroutine finishes.
+	defer coordinator.wg.Done()
+
+	// Release the semaphore after the goroutine finishes which unblocks another one.
+	defer func() { <-coordinator.sem }()
+
+	var proof types.Proof
+	// proofBz is not expected to fail unmarshalling since it is should have
+	// passed EnsureWellFormedProof validation in MsgSubmitProof handler.
+	// Panic if it fails unmarshalling.
```

Comment:
#PUC with extra details related to block halting/production.

<img width="798" alt="Screenshot 2025-01-24 at 4 34 14 PM" src="https://github.com/user-attachments/assets/896c22bd-8753-4f05-833c-998dd9c70d7c" />